[branch-jj] ODBC driver and server side fixes for working with Tableau

sumwale opened the pull request:
## Changes proposed in this pull request

- fixed native driver to use 64-bit cursor and statement handles
- correcting the type order as per snappydata.thrift
- fixing serialization of DATE, TIME and TIMESTAMP as per changes to snappydata.thrift
- additions for upstream GemFire connector
- version changes for GemFire connector so that client-server communication now uses
  underlying GFE version (8.1) while all else uses normal GFXD version (1.5.5)
- [SNAP-303] Adding non-store hive tables to JDBC/ODBC meta-data
- get column table meta-data from hive meta-data (already on master cherry-picked from here)
- Instead of having STREAM, SAMPLE and TOPK tables pose like EXTERNAL tables,
  now honour explicit type names for them respectively.
- Route query to Spark even for assertion errors by Derby parser/optimizer
- Added column value in data format errors thrown by C++ driver

## Patch testing

precheckin; ongoing for GemFire connector

## ReleaseNotes changes

For Tableau and GemFire connector support

## Other PRs 

See snappydata [branch-jj] PR

Commit:
git cherry-pick -n 4acea7fefede4ab0cc15773cac1940cf736b8da7

Use gemfirexd URL from shell in rowstore mode

Set the "client connect" URL to use jdbc:gemfirexd:// in rowstore mode (or when thrift is set to false explicitly)

Updated the C++ driver for latest thrift IDL

Updated the code to use accessor method instead of reading/updating the USE_THRIFT_AS_DEFAULT static directly (which is now private)

Commit:
Bump up the versions Store 1.5.4, Spark 2.0.3-3 and Jobserver 0.6.3-3

Commit:
SNAP 1310: Fixing outofmemory issue with string datatype (#152)

* Given values of char constants should be normalized based on column type.

Commit:
Fix for StatementStatsDUnit failure (#155)

rishitesh committed:
The internal buffer of ClientClob was getting closed before being used. Removed the close() call. (#156)

Commit:
Clean up direct ByteBuffers more aggressively

Restrict creation of direct ByteBuffers to only BLOB reads.
Writes will continue to use optimized direct path is possible.

Add a property on the connection to use direct ByteBuffers for BLOB reads (off by default now).

Commit:
Change to a new version naming policy.

snappy-sachin committed:
Fixing native driver to use 64-bit cursor and statement handles

Commit:
[NATIVE] correcting the type order as per snappydata.thrift

snappy-sachin committed:
[NATIVE] fixing serialization of DATE, TIME and TIMESTAMP

These types are no longer defined as separate structs in snappydata.thrift
rather are i64 values.

Commit:
git cherry-pick -n fa687c668d3d6a1044659ddc918af4efc6f29aec

Switch unsafe to the snappy-spark

Also make it a compile time dependency so that it gets including in client/core shadow jars

Conflicts:
	build.gradle

Commit:
git cherry-pick -n e8ba7fc0481f40b566047a491360387d72a256e8

disable "install" for top-level project with jar being disabled

Conflicts:
	build.gradle

Commit:
Relocate snappy-spark-unsafe and thrift in all shadow jars

This is to avoid conflicts with upstream versions if present in classpath at runtime.

Also changed both to provided since relocation means they should be included in product separately
and cannot depend on inclusion by snappydata-client shadow jar.

Commit:
reduced snappySparkVersion in build dependency since the updated one is not published yet

dshirish committed:
Fixing the calling convention arguments for Win64 in WinNativeCalls

Now using JNA's W32APIOptions.DEFAULT_OPTIONS instead of specific options settings
(W32APIOptions handles different platforms, type conversions etc)

snappy-sachin committed:
git cherry-pick -n 097c391da4ced499a0b83222e7096204cbd23884

Updates for ODBC API

snappy-sachin committed:
git cherry-pick -n 0ebcf52b1c4c5f44d5584b8225a1e03573e1df7e

missed in last commit

snappy-sachin committed:
git cherry-pick -n 5cc25fab9d894eb343a49726d098d4146f8d3889

Moving ClientService calls out from headers since it is internal API

snappy-sachin committed:
git cherry-pick -n a072c01d692075006d597969bfe995de572134a4

Removing check on getIsPrepStmntQuery, as it is returning wrong flag value when running select * query from Tableau.

snappy-sachin committed:
git cherry-pick -n e1fcbef6ff30a5ad68a4d0eb0060a5a735cea1b5

[Sumedhs Changes] Handling windows null character (if appneded in the end of query string). Progress ODBC Driver appends windows null character while sending sql query to server.

snappy-sachin committed:
git cherry-pick -n 928f7dda642a99c2c88b5e743b0fd0e4132fc9f6

SNAP-1357: Update ODBC Driver Installer to working state.
 Fixes: Adding missing isBeforeFirst() function from earlier commit 008815a77cec16d0446c9144cbf4e60f42af9e14.

ahshahid committed:
temp changes

ahshahid committed:
added missing file

ahshahid committed:
fixed compilation failure

snappy-sachin committed:
disable keepAlive on socket to avoid timeout errors seen in AWS/Azure setup

ahshahid committed:
Adding sumedh's change of using common Version 8.1 for Gfxd & GemFire

Commit:
[SNAP-303] Adding non-store hive tables to JDBC/ODBC meta-data (#164)

- New HiveTablesVTI which reads the information from hive meta-store and displays it in SYS.HIVETABLES
- JDBC/ODBC meta-data queries updated to do a UNION with SYS.HIVETABLES
- Added the new table types to EmbedDatabaseMetaData, ij parser and other relevant places
  (note that all types are not added at all places since those wanting to query can
   use type as "EXTERNAL TABLE" and all the non-store hive tables will be shown)
- Updated ODBCMetadataGenerator to inject MAXWIDTH column for getColumns call with SYS.HIVETABLES
- Add any new schemas introduced by hive tables in "getSchemas" meta-data query
- use common Version GFE 8.1 for GemFire connector
- avoid using CacheFactory.getAnyInstance from inside QueuedSynchronizer and couple other places;
  causes deadlock due to lock order inversion with tomcat ConnectionPool locks
  and shutdown thread already holding lock on the CacheFactory.class

Commit:
ignore exception in HiveTablesVTI when querying hive meta-store

Commit:
Get column table meta-data from hive meta-data

Commit:
Add explicit checks for STREAM, SAMPLE, TOPK tables

Instead of having STREAM, SAMPLE and TOPK tables pose like EXTERNAL tables,
now honour explicit type names for them respectively.

snappy-sachin committed:
Route query to Spark even for assertion errors by Derby parser/optimizer

Added column value in data format errors thrown by C++ driver.

Commit:
Merge remote-tracking branch 'origin/snappy/master' into branch-jj-1.5.4.1

Commit:
corrections after snappy/master merge

Commit:
correcting Version.CURRENT and Version.CURRENT_GFE uses

Commit:
correct duplicate segment in getColumns in merge

kneeraj commented on build.gradle:
```diff
@@ -39,7 +39,7 @@ allprojects {
     scalaBinaryVersion = '2.11'
     scalaVersion = scalaBinaryVersion + '.8'
-    snappySparkVersion = '2.0.2.3'
+    sparkVersion = '2.0.0'
```
Can you please explain both the build.gradle change?

sumwale replied:
snappySparkVersion is no longer used (after the removal of snappy-spark-unsafe dependency). Upstream spark is used for spark-unsafe which currently gets inherited from snappydata but ideally store should be independently buildable too.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireSparkConnectorCacheImpl.java:
```diff
@@ -0,0 +1,56 @@
+    af.setDataPolicy(DataPolicy.EMPTY);
+    UserSpecifiedRegionAttributes ra = (UserSpecifiedRegionAttributes)af.create();
+    ra.requiresPoolName = true;
+    this.setRegionAttributes(ClientRegionShortcut.PROXY.toString(), ra);
```
Just to understand.
So against this id 'ClientRegionShortcut.PROXY.toString()' we create a regionattribute object and store in the cache's map. Probably this will be used by the client regions on this vm. Would we require any other attribute? The region shortcut in the cache xml will have a set of attributes against this id. So the final region attribute will be DataPolicy.NORMAL + whatever is mentioned in the region shortcut against this id?

sumwale replied:
I suppose all usages of GemFireSparkConnectorCacheImpl will only be using PROXY shortcut. @ahshahid can you clarify?

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/cache/client/internal/AbstractOp.java:
```diff
@@ -47,7 +47,7 @@
 
   protected AbstractOp(LogWriterI18n lw, int msgType, int msgParts) {
-    this.msg = new Message(msgParts, Version.CURRENT);
+    this.msg = new Message(msgParts, Version.CURRENT_GFE);
```
So the assumption is our CURRENT_GFE Version will be compatible with their server version. Assuming they are using a newer or equal version from what we forked out?

sumwale replied:
We were compatible for at least 7.1 as part of initial rolling upgrade support effort. Perhaps our CURRENT_GFE should point to the same. As of now if 8.1 works well enough for our scenarios so will stick to that since its not clear if the newer GemFire versions will continue supporting 7.1 version (or will be tested with those) since those are now unsupported versions. We may need to added 8.1 updated classes to client if there is trouble.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/distributed/internal/tcpserver/TcpClient.java:
```diff
@@ -121,7 +121,7 @@ public static Object requestToServer(InetAddress addr, int port, Object request,
     try {
       DataOutputStream out=new DataOutputStream(sock.getOutputStream());
-      if (serverVersion < Version.CURRENT_ORDINAL) {
+      if (serverVersion < Version.CURRENT_GFE_ORDINAL) {
```
Doesn't this mean forward compatible?

sumwale replied:
This is backward compatible. If version is smaller then create a versioned data stream else a normal one.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/distributed/internal/tcpserver/TcpServer.java:
```diff
@@ -340,11 +340,11 @@ public void run() {
 
           int gossipVersion = input.readInt();
-          short versionOrdinal = Version.CURRENT_ORDINAL;
+          short versionOrdinal;
```
You have changed the default assignment? Any chances it remains unassigned.

sumwale replied:
Java will fail compilation if variable remains unassigned.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/GemFireStore.java:
```diff
@@ -1330,6 +1363,88 @@ else if (this.myKind.isAgent()) {
+
+    List<ServerLocation> prefServers = null;
+    if (servers.size() > 0) {
+      String sparkIp = System.getenv("SPARK_LOCAL_IP");
```
Who is suppossed to set this SPARK_LOCAL_IP env variable?

sumwale replied:
SPARK_LOCAL_IP is a standard spark variable to set the address Spark will bind to. Users may choose to set it. Perhaps it should be tied to our bind-address?

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/GemFireStore.java:
```diff
@@ -1330,6 +1363,88 @@ else if (this.myKind.isAgent()) {
+        hostName = "";
+      }
+      int spacing = servers.size() / 3;
+      if (spacing != 0) {
```
So spacing more than 3 servers. Can you please put some comment as to why this special consideration for 3 servers were required. May be @ahshahid did this.

sumwale replied:
This is all code added by @ahshahid for connector. Can you explain @ahshahid?

kneeraj commented on tests/core/src/main/java/com/gemstone/gemfire/distributed/internal/tcpserver/TcpServerBackwardCompatDUnitDisabledTest.java:
```diff
@@ -119,7 +119,8 @@ public void testGossipVersionBackwardCompatibility() {
       public void run2() throws CacheException {
         try {
-          TcpServer.getGossipVersionMapForTestOnly().put(TcpServer.TESTVERSION-100, Version.CURRENT_ORDINAL);
+          TcpServer.getGossipVersionMapForTestOnly().put(TcpServer.TESTVERSION-100,
```
For completeness you would have changed this. I don't think we can run this gemfire test or can we?

sumwale replied:
Yes I changed all instances of client code/tests to use GFE version.

kneeraj reviewed (commented):
LGTM. Understood all the changes and looks all ok to me. Some minor comments and some questions to verify whether the understanding is ok or not.
BTW I don't think the WAN should also get affected as the poolconnection established specially for the remote locator ds configuration should not affect the WAN client connection. They must be separate. Isn't it?

Commit:
reverting to GemFire version as 7.1 for proper GemFire compatibility

- changed version number for 7.1 to match that in released Pivotal GemFire product
- GFXD 1.5.5 now uses GFE 7.1 as underlying Gemfire version for proper compatibility with external GemFire clusters (as client)
